File: BOJ/10216.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <queue>
#include <vector>
#include <math.h>
using namespace std;
typedef pair<pair<int, int>, int> P;
P board[3005];
bool visited[3005];
int n,t;
vector<vector<int>> a;
void dfs(int here) {
	if (visited[here]) return;
	visited[here] = true;
	for (int i = 0; i < a[here].size(); i++) {
		int next = a[here][i];
		if (!visited[next]) {
			dfs(next);
		}
	}
}
int main(){
	scanf("%d", &t);
	while (t--) {
		scanf("%d", &n);
		memset(board, 0, sizeof(board));
		memset(visited, false, sizeof(visited));
		for (int i = 1; i <= n; i++) {
			scanf("%d%d%d", &board[i].first.first, &board[i].first.second, &board[i].second);
		}
		a.clear();
		a.resize(n + 1);
		for (int i = 1; i <= n; i++) {
			for (int j = i; j <= n; j++) {
				if (i == j) continue;
				else {
					int x = board[i].first.first - board[j].first.first;
					int y = board[i].first.second - board[j].first.second;
					double dist = (double)sqrt(x*x+y*y);
					double range = (double)board[i].second +(double)board[j].second;
					if (range >= dist) {
						a[i].push_back(j);
						a[j].push_back(i);
					}
				}
			}
		}
		int group = 0;
		for (int i = 1; i <= n; i++) {
			if (!visited[i]) {
				dfs(i);
				group++;
			}
		}
		printf("%d\n", group);
	}
	return 0;
}

File: BOJ(Java 공부용)/10216.java
import java.io.*;
import java.util.ArrayList;
import static java.lang.Math.sqrt;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0,1,-1,0};
    static int[] goy = {1,0,0,-1};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().split(" ")[0]);
        for(int z = 0;z<tc;z++) {
            int n = Integer.parseInt(br.readLine().split(" ")[0]);
            boolean[] visited = new boolean[n];
            ArrayList<p> arr = new ArrayList<>();
            ArrayList<Integer>[] graph = new ArrayList[n];
            for(int i=0;i<n;i++) graph[i] = new ArrayList<>();
            for(int i=0;i<n;i++) {
                String[] input = br.readLine().split(" ");
                int first = Integer.parseInt(input[0]), second = Integer.parseInt(input[1]), third = Integer.parseInt(input[2]);
                arr.add(new p(first, second, third));
            }

            for(int i=0;i<n;i++) {
                for(int j=i;j<n;j++) {
                    if(i != j) {
                        int x = arr.get(i).first - arr.get(j).first;
                        int y = arr.get(i).second - arr.get(j).second;
                        double dist = sqrt(x*x+y*y);
                        double range = (double)arr.get(i).third +(double)arr.get(j).third;
                        if(range >= dist ) {
                            graph[i].add(j);
                            graph[j].add(i);
                        }
                    }
                }
            }

            int group = 0;
            for(int i=0;i<n;i++) {
                if(!visited[i]) {
                    dfs(i,visited,graph);
                    group++;
                }
            }
            bw.write(group+"\n");
        }
        bw.flush();
        bw.close();
    }
    private static void dfs(int here,boolean[] visited, ArrayList<Integer>[] graph) {
        visited[here] = true;
        for(int i=0;i<graph[here].size();i++) {
            int next = graph[here].get(i);
            if(!visited[next]) dfs(next,visited,graph);
        }
    }
    private static class p {
        int first,second,third;
        private p(int first,int second,int third) {
            this.first = first;
            this.second = second;
            this.third = third;
        }
    }
}